### ISSUE TITLE
[BUG] 20% CPU overhead in kindness mode

### ISSUE THREAD

**Issue by Android-PowerUser** [guardianproject/orbot-android#1183]
As soon as the kindness mode is on, Orbox starts a process that uses about 20% of the CPU and cannot be turned off. Not even if the mode is turned off. I don't think that's intended. It helps to end the process or the app.

![IMG_20240920_122745](https://github.com/user-attachments/assets/88929ad3-db8c-4a0c-95e6-ba106cc5a3ee)
Xiaomi POCO X3 pro

**Comment by tladesignz** [comment]
Profiling memory and CPU usage confirms this:

The spike in CPU usage, is when I turn Kindness mode on. Then I turned it off. Then I took the screenshot. No downward curve to be witnessed.

![Image](https://github.com/user-attachments/assets/baeb14be-54e9-4200-8ff7-27b18426fd8e)


However, the code looks totally unsuspicious.

I debugged it, [`#disableSnowflakeProxy`](https://github.com/guardianproject/orbot-android/blob/master/orbotservice/src/main/java/org/torproject/android/service/OrbotService.java#L546-L560) seems to be getting called without problem. 

```java
    public synchronized void disableSnowflakeProxy() {

        if (mSnowflakeProxy != null) {
            mSnowflakeProxy.stop();
            logNotice(getString(R.string.log_notice_snowflake_proxy_disabled));

            if (Prefs.showSnowflakeProxyMessage()) {
                var message = getString(R.string.log_notice_snowflake_proxy_disabled);
                new Handler(getMainLooper()).post(() -> Toast.makeText(getApplicationContext(), message, Toast.LENGTH_LONG).show());
            }
            mSnowflakeProxy = null;
        }


    }
```

Actually a little too often for my tastes... I witnessed WiFi changes which didn't really happen - the device seemed connected continuously, but Snowflake Proxy was turned off and on again.

Anyway, a logic bug in Orbot itself seems out of the question.


But, [IPtProxy looks good, too](https://github.com/tladesignz/IPtProxy/blob/master/IPtProxy.go/kindness.go#L90-L97):

```go
// Stop - Stop the Snowflake proxy.
func (sp *SnowflakeProxy) Stop() {
	if sp.isRunning {
		sp.proxy.Stop()
		sp.proxy.EventDispatcher.RemoveSnowflakeEventListener(sp)
		sp.isRunning = false
		sp.proxy = nil
	}
}
```

And the [Snowflake code](https://gitlab.torproject.org/tpo/anti-censorship/pluggable-transports/snowflake/-/blame/main/proxy/lib/snowflake.go?ref_type=heads#L817-833), just as well:

```go
	for ; true; <-ticker.C {
		select {
		case <-sf.shutdown:
			return nil
		default:
			tokens.get()
			sessionID := genSessionID()
			sf.runSession(sessionID)
		}
	}
	return nil
}

// Stop closes all existing connections and shuts down the Snowflake.
func (sf *SnowflakeProxy) Stop() {
	close(sf.shutdown)
}
```


So… I'm at a loss, here.

@cohosh, sorry that I have to summon you once again… Do you have any ideas?

**Comment by syphyr** [comment]
How about this:
https://github.com/guardianproject/orbot-android/issues/1241

**Comment by cohosh** [comment]
Hmm, I'll poke around a bit. I wonder if there is a race condition with the shutdown channel getting closed, and the proxy also waiting for a new snowflake.

**Comment by Android-PowerUser** [comment]
I seem to remember seeing the 20% once when I hadn't turned on the Kindness mode yet. But it usually turns on when the mode is turned on.

**Comment by cohosh** [comment]
@Android-PowerUser what version of orbot are you using?

**Comment by Android-PowerUser** [comment]
Version 17.3.2-RC-1-tor-0.4.8.12

**Comment by WofWca** [comment]
Seems to be related to https://github.com/guardianproject/orbot-android/issues/1091.

**Comment by 8lightning** [comment]
Each month, I have nearly a hundred gigabytes of unused bandwidth. Rather than letting it go to waste, I would like to donate it, and Snowflake appears to be the most suitable relay for my network.
Currently, I run Snowflake on my PC as a browser extension, but the system is not online 24/7. For that reason, I am considering running it on my Android phone instead. However, using Orbot with “Kindness Mode” enabled causes significant battery drain and excessive heat generation on the device.
I hope this issue can be addressed and resolved soon.

**Comment by bitmold** [comment]
It's possible that @munzzyy's theory in #1807 may partially explain this CPU overhead:

> The stopped proxy that keeps running (https://github.com/guardianproject/orbot-android/issues/1183). enableProxy() takes its lock, then does the real work in a coroutine that has no lock at all. A stopProxy() that lands before that coroutine runs sees proxy == null, does nothing, and the coroutine then starts a proxy nothing owns. The service stops and restarts the proxy on every network callback, so this window gets hit in normal use, and an orphaned proxy is a plausible source of the CPU drain reported there. Starts now carry a generation number: if a stop or a newer start won while the ports were being mapped, the late start closes its own mappings and never calls start().

> There is a second copy of the same race one layer down. IPtProxy's Start() flips its isRunning flag inside a goroutine it spawns, and its Stop() is a no-op until that flag is set, so a fast stop is silently ignored and the proxy comes up anyway:

^ this may explain earlier comments of still seeing CPU usage after disabling snowflake proxy ...

**Comment by munzzyy** [comment]
That maps cleanly to the "cannot be turned off" part. When the stop loses the race, the proxy starts with nothing holding a reference, so turning kindness mode back off has nothing to stop and it runs until the app is killed, which is what Android-PowerUser and 8lightning describe. The IPtProxy goroutine race is the same shape one layer down, so disabling snowflake on its own can leave it up too. #1807's generation check closes both windows. I have not reproduced the 20% on a device, so I would not call it the whole story, but an orphaned proxy is a concrete cause for the part where turning it off does nothing.

**Comment by tladesignz** [comment]
Well, one would need to test this assumption with debug halts at the start and stop code, to see, if stop is really called before start finished.

**Comment by munzzyy** [comment]
Agreed that debug halts on a device would confirm it. The window is not only an assumption though: enableProxy releases its lock before the coroutine that actually creates the proxy runs, so a stopProxy in that gap sees proxy == null and no-ops, and the coroutine then brings one up that nothing owns. The service re-runs stop and start on every network callback, so that interleaving is reachable in normal use rather than a rare timing fluke. Halts at start and stop would pin down how often it lands and confirm it is the source of the 20%.
